### PR TITLE
Export optimizations

### PR DESCRIPTION
**Pull request by wschnepp**
Optimizations & Fixes to make the gltf export go faster & correcterer.

### CODE REVIEW

File: Runtime/Scripts/GLTFRecorderComponent.cs [hunk 1]
```diff
@@ -34,12 +34,7 @@ public class GLTFRecorderComponent : MonoBehaviour
 	    public UnityEvent recordingStarted;
 		public UnityEvent<string> recordingEnded;
 
-		private double CurrentTime =>
-#if UNITY_2020_1_OR_NEWER
-			Time.timeAsDouble;
-#else
-			Time.time;
-#endif
+		private float CurrentTime => Time.time;
```
Comment (Eregerog) [inline Runtime/Scripts/GLTFRecorderComponent.cs:37]: Don't we need double-precision? In dProB, that is necessary for projects to be longer than 3+ months

File: Runtime/Scripts/GLTFSceneExporter.cs [hunk 2]
```diff
@@ -101,10 +108,22 @@ public override bool BeforeMaterialExport(GLTFSceneExporter exporter, GLTFRoot g
...
+	{
+		void AddAnimationData(
+			Object animatedObject,
+			string propertyName,
+			GLTFAnimation animation,
+			AnimationInterpolationType interpolationType,
+			float[] times,
+			object[] values
+		);
+	}
```
Comment (Eregerog) [inline Runtime/Scripts/GLTFSceneExporter.cs]: Document: What is this interface used for, when would I want to call the method?

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Unity.Profiling;
+
+namespace UnityGLTF
+{
+	// Contains some util functions to remove duplicate entries from the keyframe arrays
+    internal static class AnimationFilteringUtils
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]: Indentation is weird here

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			    // cache-friendly optimized loop that only does one comparison+one AND per iteration,
+			    // re-using the last result.
+			    for (var i = 0; i < values.Count - 1; i++) {
+				    var nextEquals = values[i].Equals(values[i + 1]);
+				    if (lastEquals && nextEquals) {
+					    foundDuplicates.Enqueue(i);
+				    }
+				    lastEquals = nextEquals;
+			    }
+				removeAnimationUnneededKeyframesFindDuplicatesMarker.End();
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]: indentation is also weird here

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+			    if (foundDuplicates.Count <= 0) return (times, values);
+			    removeAnimationUnneededKeyframesCopyWithoutDuplicatesMarker.Begin();
+			    var t2 = new List<float>(times.Count);
+			    var v2 = new List<object>(values.Count);
+			    
+			    var nextDuplicate = foundDuplicates.Dequeue();
+			    for (var i = 0; i < values.Count; i++) {
+				    if (i == nextDuplicate) {
+					    // deque may fail due to there not being any more duplicates - we dont want
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]:
```suggestion
					    // dequeue may fail due to there not being any more duplicates - we dont want
```

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+					    // to do anything different in that case since we still need to copy over
+					    // the remaining values to the new lists.
+					    // But we need to handle the case that the deque fails - so use TryDequeue.
+					    // Deque into a separate variable & assign on success to make sure an
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]:
```suggestion
					    // Dequeue into a separate variable & assign on success to make sure an
```

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+					    // unexpected integer return value when deque failed cannot break the logic.
+					    if (foundDuplicates.TryDequeue(out var n)) nextDuplicate = n;
+					    continue;
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]: This code would be easier to read with a if / else instead of a continue

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+			    equals &=
+				    (lastExportedSectionStart >= prevSectionStart
+					    || array[lastExportedSectionStart + i].Equals(array[sectionStart + i]))
+				    && array[prevSectionStart + i].Equals(array[sectionStart + i])
+				    && array[sectionStart + i].Equals(array[nextSectionStart + i]);
+			    if (!equals) return false;
+		    }
+
+		    return true;
+	    }
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs:154]: I am quite confused by this method. For someting so general, like checking if two ranges in an array have equal values (which I guess this method does), the parameters have weird names like "export" that do not make sense in the context of such a generic method.

Either better parameter names or a comment would help

File: Runtime/Scripts/SceneExporter/ExporterAnimation.cs [hunk 9]
```diff
@@ -1328,7 +1328,7 @@ private bool BakePropertyAnimation(PropertyCurve prop, float length, float bakin
 			var nbSamples = Mathf.Max(1, Mathf.CeilToInt(length * bakingFramerate));
 			var deltaTime = length / nbSamples;
 
-			var _times = new List<double>(nbSamples * 2);
+			var _times = new List<float>(nbSamples * 2);
```
Comment (Eregerog) [inline Runtime/Scripts/SceneExporter/ExporterAnimation.cs:1331]: As before, is this fine?

File: Runtime/Scripts/Timeline/Internal/Samplers/VisibilitySampler.cs [hunk 10]
```diff
@@ -6,23 +7,27 @@ internal sealed class VisibilitySampler : AnimationSampler<GameObject, bool>
     {
         public override string PropertyName => "visibility";
 
-        internal VisibilityTrack startNewAnimationTrackAtStartOfTime(AnimationData data, double time) =>
+        public override AnimationInterpolationType InterpolationType => AnimationInterpolationType.STEP;
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/Samplers/VisibilitySampler.cs:14]: I guess this leads to invisible objects no longer expanding / contracting?

File: Tests/Editor/RecorderTests.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+using UnityGLTF.Timeline.Samplers;
+
+namespace Tests.Editor
+{
+    public class RecorderTests
+    {
+        // Tests for testing that merging visibility and scale tracks works correctly
+
+		private AnimationTrack<GameObject, bool>? visibilityTrack;
+		private AnimationTrack<Transform, Vector3>? scaleTrack;
```
Comment (Eregerog) [inline Tests/Editor/RecorderTests.cs:15]: indentation is weird again

File: Runtime/Scripts/Timeline/Internal/AnimationTrack.cs [hunk 12]
```diff
@@ -2,69 +2,129 @@
...
+        Object? AnimatedObjectUntyped { get; }
         string PropertyName { get; }
-        double[] Times { get; }
-        object[] Values { get; }
+     
+        AnimationInterpolationType InterpolationType { get; }
         
-        double? LastTime { get; }
-        object? LastValue { get; }
+        float[] Times { get; }
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/AnimationTrack.cs:18]: Is this change from double to float really fine?

File: Runtime/Scripts/Timeline/Internal/AnimationTrack.cs [hunk 13]
```diff
@@ -2,69 +2,129 @@
...
+            // Consider the following example:
+            // Event 1: Recording sample at time t1 with value v1, samples = [(t1, v1)], lastSample = (t1, v1), secondToLastSample = null
+            // Event 2: Recording sample at time t2 with value v2, samples = [(t1, v1), (t2,v2)], lastSample = (t2, v2), secondToLastSample = (t1, v1)
+            // Event 3: Recording sample at time t2 with value v2, samples = [(t1, v1), (t2,v2)], lastSample = (t2, v2), secondToLastSample = (t2, v2)
+            //          the logic is now in an unexpected state, where it thinks the (t2,v2) sample is not necessary
+            // Event 4: Recording sample at time t3 with value v2, samples = [(t1, v1), (t3,v2)], lastSample = (t3, v2), secondToLastSample = (t2, v2).
+            // If you are wondering, why this can even happen since duplicate timestamps are already checked for in the GLTFRecorder:
+            // Yeah i am not sure why either, but it definitely happens!
+            // Finding this bug cost me 3 days of my life, please don't remove this check.
+            if (time <= LastTime) {
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/AnimationTrack.cs:116]: Good explanation, and horrible that this is even necessary 😭 (you needn't change anything, though)

File: Runtime/Scripts/AssemblyInfo.cs [hunk 14]
```diff
@@ -1,5 +1,5 @@
-﻿
-
-using System.Runtime.CompilerServices;
+﻿using System.Runtime.CompilerServices;
 
 [assembly: InternalsVisibleTo("UnityGLTFEditor")]
+[assembly: InternalsVisibleTo("UnityGLTFTests.Editor")]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
```
Comment (Eregerog) [inline Runtime/Scripts/AssemblyInfo.cs]: Comment: Why is this needed?

File: Runtime/Scripts/Timeline/Internal/VisibilityTrackMergeHelpers.cs [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+                var visTime = currentVisibilityTime;
+                var scaleTime = currentScaleTime;
+                var visible = currentVisibility;
+                var scale = currentScale;
+
+                if (visTime.nearlyEqual(scaleTime)) {
+                    
+                    if (visTime <= 0) {
+                        yield return (visTime, visible ? (lastScale ?? scale) : Vector3.zero);
+                    }
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/VisibilityTrackMergeHelpers.cs:54]: Why is this check here? I'm a bit confused

Comment (Eregerog) [reply Runtime/Scripts/Timeline/Internal/VisibilityTrackMergeHelpers.cs:54]: This goes for all <= checks - i guess they are here for initializing at time 0?

**Review by Eregerog — CHANGES_REQUESTED** [review]

File: Runtime/Scripts/GLTFRecorderComponent.cs (hunk 1 above)
Comment (wschnepp) [reply Runtime/Scripts/GLTFRecorderComponent.cs:37]: gltf cant represent that.
The animation timestamp _is always_ a single precision float, a hard requirement in the gltf spec,
so it makes absolutely zero sense to fake double support in the plugin only to then break the exported gltf by converting the double to float and ending up with the same timestamp multiple times, which is a violation of the gltf spec

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs (hunk 3 above)
Comment (wschnepp) [reply Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]: send help
image

File: Runtime/Scripts/Timeline/Internal/VisibilityTrackMergeHelpers.cs (hunk 15 above)
Comment (wschnepp) [reply Runtime/Scripts/Timeline/Internal/VisibilityTrackMergeHelpers.cs:54]: yes basically that - just safety to make sure the value at time 0 is set to a sensible value

File: Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs (hunk 3 above)
Comment (wschnepp) [reply Runtime/Scripts/SceneExporter/AnimationFilteringUtils.cs]: i think i fixed these, but not sure since rider shows this properly

File: Runtime/Scripts/Timeline/GLTFRecorder.cs [hunk 16]
```diff
@@ -269,148 +343,143 @@ private static void SafeDestroy(Object obj)
...
+		private void collectAndProcessSingleTrack(
+			AnimationDataCollector gltfSceneExporter,
+			GLTFAnimation animation,
+			AnimationTrack track,
 			VisibilityTrack? visibilityTrack,
-			BaseAnimationTrack<Transform, Vector3>? scaleTrack
+			bool calculateTranslationBounds,
+			ref bool foundFirstTranslationTrack,
+			ref Bounds translationBounds,
+			ref bool foundScaleTrack
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/GLTFRecorder.cs:394]: I'm not a big fan of reafs, are these really needed?

**Review by Eregerog — APPROVED** [review]

File: Runtime/Scripts/Timeline/GLTFRecorder.cs (hunk 16 above)
Comment (wschnepp) [reply Runtime/Scripts/Timeline/GLTFRecorder.cs:394]: in this case they are kind of needed - or i would have to return these three each call
